presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CombineApproxPercentileFunctions.java

```diff
@@ -0,0 +1,289 @@
+import java.util.Set;
+
+import static com.facebook.presto.SystemSessionProperties.isCombineApproxPercentileEnabled;
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.DoubleType.DOUBLE;
+import static com.facebook.presto.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * For multiple approx_percentile() function calls on the same column with different percentile arguments, combine them to one call on an array of percentile arguments.
+ * <p>
+ * From:
+ * <pre>
+ * - Aggregation (approx_percentile(col, 0.2), approx_percentile(col, 0.8)
+ * </pre>
+ * To:
+ * <pre>
+ * - Project (approx_percentile_results[1], approx_percentile_results[2])
+ *   - Aggregation (approx_percentile_results <- approx_percentile(col, array)
+ *     - Project (col <- col, array <- [0.2, 0.8])
+ * </pre>
+ * <p>
+ */
+
+public class CombineApproxPercentileFunctions
+        implements Rule<AggregationNode>
+{
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public CombineApproxPercentileFunctions(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(CombineApproxPercentileFunctions::hasMultipleApproxPercentile);
+
+    private static boolean hasMultipleApproxPercentile(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream().filter(agg -> agg.getCall().getDisplayName().equals("approx_percentile")).count() > 1;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isCombineApproxPercentileEnabled(session);
+    }
+
+    private static int getPercentileArgIdx(List<RowExpression> arguments)
+    {
+        // approx_percentile(x, percentage) -> arguments.size() == 2
+        // approx_percentile(x, percentage, accuracy) -> arguments.size() == 3 && arguments.get(1) is Double
+        // approx_percentile(x, w, percentage) -> arguments.size() == 3 && arguments.get(1) is BigInt
+        // approx_percentile(x, w, percentage, accuracy) -> arguments.size() == 4
+        if (arguments.size() == 2 || (arguments.size() == 3 && (arguments.get(1).getType() instanceof DoubleType))) {
+            return 1;
+        }
+        else if (arguments.size() == 4 || (arguments.size() == 3 && (arguments.get(1).getType() instanceof BigintType))) {
+            return 2;
+        }
+        return -1;
```

Comment:
Can we use the function resolution to do this in a better way?

Reply:
Changed to use function handle to check the position of percentile arguments.